### ISSUE TITLE
Optional-of-optional not supported

### ISSUE THREAD

**Issue by jfancher** [immutables/immutables#483]
Perhaps it's a marginal use case, and I'm not sure there is a good solution to it. But I thought I'd put it up here for discussion.

````java
@Value.Immutable
interface Foo {
  Optional<Optional<String>> getBar();
}
````

This doesn't compile because it tries to generate `bar(Optional<String>)` and `bar(Optional<Optional<String>>)` in the builder, which have the same (erased) signature (similarly with `withBar`).

**Comment by elucash** [comment]
thank you, maybe there will be some way to handle it or warn about it

**Comment by bartosz-bilicki** [comment]
why you ever need Optional<Optional<String>> ?

I think that Optional<Optional<"abc">> is not diffrent from Optional<"abc">. Which is not diffrent from "abc" (I mean business reason and actual, business and logical meaning, not Java types compatibility).

About Optional<Optional.empty()> - how it's diffrent from just Optional.empty() ?

It is true that sometimes stream API may return Optional<Optional<String>>, but it is usually wrong ussage of API. If you use flatMap instead of map, API will return Optional<String>.

Going further, one may invent Optional<Otional<Optional<Optional<String>>>> and even more.

**Comment by elucash** [comment]
I would definitely agree that we don't have to "support" this case, but we could rather detect `Optional<Optional<..` this and treat it as "nothing-special" attribute, in this way convenience overloads will not get in the way.

**Comment by bartosz-bilicki** [comment]
Idea that Immutables should work with Optional\<Optional\<>> is reasonable.

Note that java.util.Optional is meant as 'limited mechanism for library method return types' (http://stackoverflow.com/a/26328555/1849837) in first place. Therefore using optional as field type or parameter is abusing original idea. Nevertheless, it works (compiles).

Following the same logic, I believe that Immutables should support  Optional\<Optional\< even it does not look right.

**Comment by jfancher** [comment]
> why you ever need Optional<Optional<String>> ?

The code I wanted to write was similar to this:

````java
// a model object
interface SomeThing {
  String getName(); // required property
  Optional<String> getDescription(); // optional property
}
````

````java
// describes a partial update to the above model type
interface SomeThingUpdate {
  // empty = leave it alone
  // set = change to that value
  Optional<String> getName();

  // empty = leave it alone
  // set-with-empty = clear it
  // set-with-something = change to that value
  Optional<Optional<String>> getDescription();
}
````

You may question this design, but I've found it to be a sensible and useful pattern for this project. In particular, the awkward part is actually calling `update.getDescription()`, which is only really intended to be done in one place -- creation has a smoother API (see below), and that's the more important (more widely used) piece.

As far as the intended usage of optionals, two notes:

First, this pattern doesn't necessarily require any optionals as parameters (it would require a field, but that's completely hidden in generated code). The builder would look something like:

````java
interface SomeThingUpdate {
  interface Builder {
    Builder clearDescription();
    Builder setDescription(String description);
  }
}
````

I wouldn't expect the Immutables library to generate this -- I would put a cover over it. Just saying that API-wise two-level `Optional` does not imply `Optional` parameters.

Second, if we're going by Goetz, `Optional` shouldn't be in Immutables at all:
> I think routinely using it as a return value for getters would definitely be over-use.

I disagree with this, and I think the Immutables library devs implicitly do as well by the support that's already there. Or rather, it's not that I disagree with his basic point, but I think "using null for such was overwhelmingly likely to cause errors" applies to domain model getters more often than not.

Anyway, there are easy work-arounds, but as @elucash says it doesn't seem awful to just detect the conflict and do something simple to get around it.

**Comment by elucash** [comment]
Optional of Optional will be treated as regular attribute (without special initializers) and there's warning about this (can be suppressed by `@SuppressWarning("immutables")`). To be released in upcoming 2.3.9

**Comment by elucash** [comment]
released in 2.3.9
